File: src/frontend/src/api/backup.ts
/** Admin: backup + restore API.
 *
 * Multipart helpers use raw fetch because request<T> sets
 * Content-Type: application/json which corrupts multipart FormData;
 * blob downloads also bypass JSON parsing. Auth headers come from the shared
 * tokenStore so the auth layer stays consolidated.
 *
 * Password handling: when no password is set we GET the plain zip — that
 * matches the documented "pull from elsewhere" workflow that lives on the
 * API page. When a password is set we POST it in the body so it does not
 * leak into proxy/access logs or browser history.
 */

import { buildAuthHeaders } from "../shared/tokenStore";

const BASE = "/api";

export type DownloadProgress = {
  /** Bytes received so far. */
  loaded: number;
  /** Total bytes if the server set Content-Length, else null. */
  total: number | null;
};

async function consumeStream(
  res: Response,
  onProgress?: (p: DownloadProgress) => void,
): Promise<Blob> {
  const totalHeader = res.headers.get("Content-Length");
  const total = totalHeader ? Number(totalHeader) : null;
  if (!res.body || !onProgress) {
    return await res.blob();
  }
  const reader = res.body.getReader();
  const chunks: BlobPart[] = [];
  let loaded = 0;
  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    if (value) {
      chunks.push(value);
      loaded += value.byteLength;
      onProgress({ loaded, total });
    }
  }
  const type = res.headers.get("Content-Type") ?? "application/octet-stream";
  return new Blob(chunks, { type });
}

export async function downloadBackup(
  password?: string,
  onProgress?: (p: DownloadProgress) => void,
): Promise<Blob> {
  let res: Response;
  if (password) {
    const fd = new FormData();
    fd.append("password", password);
    res = await fetch(`${BASE}/admin/backup/download`, {
      method: "POST",
      body: fd,
      headers: buildAuthHeaders(),
    });
  } else {
    res = await fetch(`${BASE}/admin/backup`, {
      headers: buildAuthHeaders(),
    });
  }
  if (!res.ok) {
    throw new Error(`API ${res.status}: ${await res.text()}`);
  }
  return await consumeStream(res, onProgress);
}

export type BackupConflict = {
  table: string;
  natural_key: string | null;
  incoming: Record<string, unknown>;
  existing: Record<string, unknown>;
};

export type BackupInspectionReport = {
  format_version: number;
  exported_at: string | null;
  encrypted: boolean;
  table_counts: Record<string, number>;
  conflicts: BackupConflict[];
};

export async function inspectBackup(
  file: File,
  password?: string,
): Promise<BackupInspectionReport> {
  const fd = new FormData();
  fd.append("file", file);
  if (password) fd.append("password", password);
  const res = await fetch(`${BASE}/admin/backup/inspect`, {
    method: "POST",
    body: fd,
    headers: buildAuthHeaders(),
  });
  if (!res.ok) {
    throw new Error(`API ${res.status}: ${await res.text()}`);
  }
  return (await res.json()) as BackupInspectionReport;
}

export type BackupRestoreResult = {
  inserted: number;
  skipped: number;
  overwritten: number;
};

export async function restoreBackup(args: {
  file: File;
  password?: string;
  mode: "fresh" | "addon";
  resolutions?: Record<string, "skip" | "overwrite">;
}): Promise<BackupRestoreResult> {
  const fd = new FormData();
  fd.append("file", args.file);
  fd.append("mode", args.mode);
  if (args.password) fd.append("password", args.password);
  if (args.resolutions) {
    fd.append("resolutions_json", JSON.stringify(args.resolutions));
  }
  const res = await fetch(`${BASE}/admin/backup/restore`, {
    method: "POST",
    body: fd,
    headers: buildAuthHeaders(),
  });
  if (!res.ok) {
    throw new Error(`API ${res.status}: ${await res.text()}`);
  }
  return (await res.json()) as BackupRestoreResult;
}

File: src/backend/app/routers/backup.py
"""Admin backup/restore endpoints. Streams the zipped DB+media as one download."""

from __future__ import annotations

import contextlib
import json
import os
import tempfile
from datetime import datetime
from typing import Annotated

from fastapi import APIRouter, BackgroundTasks, File, Form, HTTPException, UploadFile
from fastapi.responses import FileResponse

from app.auth import AdminDep
from app.db import SessionDep
from app.services.backup_service import (
    BackupAuthError,
    BackupFormatError,
    export_backup_to_file,
    inspect_backup,
    restore_backup,
)

router = APIRouter(prefix="/admin/backup", tags=["admin-backup"])


def _filename(encrypted: bool) -> str:
    timestamp = datetime.now().strftime("%Y%m%d-%H%M")
    suffix = "-encrypted.bin" if encrypted else ".zip"
    return f"nodus-backup-{timestamp}{suffix}"


def _stream_response(
    session: SessionDep,
    background: BackgroundTasks,
    password: str | None,
) -> FileResponse:
    """Build the backup on disk and return it as a streaming FileResponse.

    The tempfile is unlinked after the response finishes via BackgroundTasks
    so we don't pile up half-finished exports.
    """
    fd, tmp_path = tempfile.mkstemp(prefix="nodus-backup-", suffix=".bin")
    os.close(fd)
    try:
        export_backup_to_file(session, tmp_path, password=password or None)
    except Exception:
        with contextlib.suppress(OSError):
            os.unlink(tmp_path)
        raise
    background.add_task(os.unlink, tmp_path)
    encrypted = bool(password)
    return FileResponse(
        tmp_path,
        media_type="application/octet-stream" if encrypted else "application/zip",
        filename=_filename(encrypted),
    )


@router.get("")
def download_backup(
    session: SessionDep,
    _admin: AdminDep,
    background: BackgroundTasks,
) -> FileResponse:
    """Download the full DB+media as a plain zip.

    Designed for scripted pulls — bearer token in the Authorization header
    is the only secret needed. No query parameters; no encryption layer.
    Use POST /api/admin/backup/download when an encryption envelope is
    required.
    """
    return _stream_response(session, background, password=None)


@router.post("/download")
async def download_backup_encrypted(
    session: SessionDep,
    _admin: AdminDep,
    background: BackgroundTasks,
    password: Annotated[str | None, Form()] = None,
) -> FileResponse:
    """Download the backup with an optional AES-256-GCM envelope.

    Identical to GET / when `password` is blank. The password is read from
    the request body so it does not leak into access logs or browser history
    the way a GET query parameter would.
    """
    return _stream_response(session, background, password=password)


@router.post("/inspect")
async def inspect_uploaded(
    session: SessionDep,
    _admin: AdminDep,
    file: Annotated[UploadFile, File(...)],
    password: Annotated[str | None, Form()] = None,
) -> dict[str, object]:
    """Read the manifest + identify which rows would conflict on add-on import."""
    payload = await file.read()
    try:
        report = inspect_backup(session, payload, password=password or None)
    except BackupAuthError as exc:
        raise HTTPException(status_code=401, detail=str(exc)) from exc
    except BackupFormatError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    return {
        "format_version": report.format_version,
        "exported_at": report.exported_at,
        "encrypted": report.encrypted,
        "table_counts": report.table_counts,
        "conflicts": [
            {
                "table": c.table,
                "natural_key": c.natural_key,
                "incoming": c.incoming,
                "existing": c.existing,
            }
            for c in report.conflicts
        ],
    }


@router.post("/restore")
async def restore_uploaded(
    session: SessionDep,
    _admin: AdminDep,
    file: Annotated[UploadFile, File(...)],
    mode: Annotated[str, Form(...)] = "addon",
    password: Annotated[str | None, Form()] = None,
    resolutions_json: Annotated[str | None, Form()] = None,
) -> dict[str, int]:
    """Apply the backup. `mode` = "fresh" wipes existing rows first; "addon" merges.

    `resolutions_json` is an optional JSON-encoded map keyed by `<table>:<key>`
    with values "skip" | "overwrite" — drives per-row conflict handling in
    addon mode. Anything missing defaults to "skip".
    """
    if mode not in {"fresh", "addon"}:
        raise HTTPException(status_code=400, detail="mode must be 'fresh' or 'addon'")
    payload = await file.read()
    resolutions: dict[str, str] = {}
    if resolutions_json:
        try:
            parsed = json.loads(resolutions_json)
            if isinstance(parsed, dict):
                resolutions = {
                    str(k): str(v)
                    for k, v in parsed.items()
                    if isinstance(v, str) and v in {"skip", "overwrite"}
                }
        except json.JSONDecodeError:
            raise HTTPException(
                status_code=400, detail="resolutions_json is not valid JSON"
            ) from None
    try:
        return restore_backup(
            session,
            payload,
            password=password or None,
            mode=mode,
            resolutions=resolutions,
        )
    except BackupAuthError as exc:
        raise HTTPException(status_code=401, detail=str(exc)) from exc
    except BackupFormatError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
